File: backend/src/services/taskService.js
const { ScrapingTask } = require('../models');

async function generateTaskId() {
  const lastTask = await ScrapingTask.findOne({
    order: [['id', 'DESC']],
  });
  const nextNum = lastTask && lastTask.id ? lastTask.id + 1 : 1;
  const padded = String(nextNum).padStart(6, '0');
  return `TASK-${padded}`;
}

async function updateTaskStatus(taskDbId, status, extraFields = {}) {
  const updateData = { status, ...extraFields };
  if (['COMPLETED', 'FAILED', 'CANCELLED'].includes(status) && !updateData.completed_at) {
    updateData.completed_at = new Date();
  }
  await ScrapingTask.update(updateData, {
    where: { id: taskDbId },
  });
}

async function incrementTaskCounter(taskDbId, field, amount = 1) {
  const task = await ScrapingTask.findByPk(taskDbId);
  if (task) {
    const current = task[field] || 0;
    task[field] = current + amount;
    await task.save();
  }
}

module.exports = {
  generateTaskId,
  updateTaskStatus,
  incrementTaskCounter,
};

File: backend/app/services/task_service.py
from sqlalchemy.orm import Session
from app.models.scraping_task import ScrapingTask
from datetime import datetime


def generate_task_id(db: Session) -> str:
    """Generate a unique sequential task ID like TASK-000001."""
    max_task = db.query(ScrapingTask).order_by(ScrapingTask.id.desc()).first()
    if max_task and max_task.id:
        next_num = max_task.id + 1
    else:
        next_num = 1
    return f"TASK-{next_num:06d}"


def get_task_by_task_id(db: Session, task_id: str, user_id: int) -> ScrapingTask:
    return (
        db.query(ScrapingTask)
        .filter(ScrapingTask.task_id == task_id, ScrapingTask.user_id == user_id)
        .first()
    )


def update_task_status(db: Session, task_db_id: int, status: str, **kwargs):
    task = db.query(ScrapingTask).filter(ScrapingTask.id == task_db_id).first()
    if task:
        task.status = status
        for key, value in kwargs.items():
            if hasattr(task, key):
                setattr(task, key, value)
        if status in ("COMPLETED", "FAILED", "CANCELLED") and not task.completed_at:
            task.completed_at = datetime.utcnow()
        db.commit()


def increment_task_counter(db: Session, task_db_id: int, field: str, amount: int = 1):
    task = db.query(ScrapingTask).filter(ScrapingTask.id == task_db_id).first()
    if task and hasattr(task, field):
        current = getattr(task, field) or 0
        setattr(task, field, current + amount)
        db.commit()
